### PR TITLE
KAFKA-16898: move TimeIndexTest and TransactionIndexTest to storage module

### PR DESCRIPTION
**Pull request by FrankYang0529**
### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Unit test for time index.
+ */
+public class TimeIndexTest {
+    private TimeIndex idx;
+    private int maxEntries = 30;
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java]: could you please add `final`?

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    private long baseOffset = 45L;
+
+    @BeforeEach
+    public void setup() {
+        this.idx = assertDoesNotThrow(() -> new TimeIndex(nonExistentTempFile(), baseOffset, maxEntries * 12));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java]: Could we initialize this variable in construction?

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class TransactionIndexTest {
+    private File file;
+    private TransactionIndex index;
+
+    @BeforeEach
+    public void setup() {
+        file = assertDoesNotThrow(() -> TestUtils.tempFile());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: Could we create them in construction?

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+/**
+ * Unit test for time index.
+ */
+public class TimeIndexTest {
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java:39]: Could you add unit test for following methods: `isFull`, `lastEntry`, `resize`, `entrySize` and `parseEntry`

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+
+public class TransactionIndexTest {
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java:38]: Could you please add test for following methods? `flush` and `deleteIfExists`?

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for this PR!

**Comment by FrankYang0529** [comment]
Hi @chia7712, thanks for the review. I address all comments. Thank you.

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    private final File file = assertDoesNotThrow(() -> TestUtils.tempFile());
+    private final TransactionIndex index = assertDoesNotThrow(() -> new TransactionIndex(0, file));
+
+    @AfterEach
+    public void teardown() {
+        assertDoesNotThrow(() -> index.close());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: We don't need to use `assertDoesNotThrow` here, as it is ok to propagate the exception

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+    @Test
+    public void testSanityCheck() throws IOException {
+        List<AbortedTxn> abortedTxns = Arrays.asList(
+                new AbortedTxn(0L, 0, 10, 11),
+                new AbortedTxn(1L, 5, 15, 13),
+                new AbortedTxn(2L, 18, 35, 25),
+                new AbortedTxn(3L, 32, 50, 40));
+        abortedTxns.forEach(txn -> assertDoesNotThrow(() -> index.append(txn)));
+        assertDoesNotThrow(() -> index.close());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+
+    @Test
+    public void testPositionSetCorrectlyWhenOpened() {
+        List<AbortedTxn> abortedTxns = new ArrayList<>(Arrays.asList(
+                new AbortedTxn(0L, 0, 10, 11),
+                new AbortedTxn(1L, 5, 15, 13),
+                new AbortedTxn(2L, 18, 35, 25),
+                new AbortedTxn(3L, 32, 50, 40)));
+        abortedTxns.forEach(txn -> assertDoesNotThrow(() -> index.append(txn)));
+        assertDoesNotThrow(() -> index.close());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+
+        TransactionIndex reopenedIndex = assertDoesNotThrow(() -> new TransactionIndex(0L, file));
+        AbortedTxn anotherAbortedTxn = new AbortedTxn(3L, 50, 60, 55);
+        assertDoesNotThrow(() -> reopenedIndex.append(anotherAbortedTxn));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+    @Test
+    public void testLastOffsetMustIncrease() {
+        assertDoesNotThrow(() -> index.append(new AbortedTxn(1L, 5, 15, 13)));
+        assertThrows(IllegalArgumentException.class, () -> index.append(new AbortedTxn(0L, 0,
+                15, 11)));
+    }
+
+    @Test
+    public void testLastOffsetCannotDecrease() {
+        assertDoesNotThrow(() -> index.append(new AbortedTxn(1L, 5, 15, 13)));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+        assertDoesNotThrow(() -> testIndex.append(new AbortedTxn(0L, 0, 10, 2)));
+        assertDoesNotThrow(testIndex::flush);
+        assertNotEquals(0, testIndex.file().length());
+        assertDoesNotThrow(testIndex::close);
+    }
+
+    @Test
+    public void testDeleteIfExists() {
+        assertTrue(file.exists());
+        assertDoesNotThrow(index::deleteIfExists);
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto

**Review by chia7712 — COMMENTED** [review]
@FrankYang0529 thanks for updated PR

**Comment by FrankYang0529** [comment]
Hi @chia7712, I remove `assertDoesNotThrow` for function call, but keep it if it's in lambda. Thanks for the suggestion.

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+        assertNotEquals(tmpParentDir, index.file().getParentFile());
+        index.updateParentDir(tmpParentDir);
+        assertEquals(tmpParentDir, index.file().getParentFile());
+    }
+
+    @Test
+    public void testFlush() {
+        File nonExistentFile = assertDoesNotThrow(() -> TestUtils.tempFile());
+        assertTrue(nonExistentFile.delete());
+        TransactionIndex testIndex = assertDoesNotThrow(() -> new TransactionIndex(0, nonExistentFile));
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: Could we remove `assertDoesNotThrow` and use try-resource to rewrite it?

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+        idx.maybeAppend(10000L, 1000L, true);
+    }
+
+    @Test
+    public void testSanityCheck() throws IOException {
+        idx.sanityCheck();
+        appendEntries(5);
+        TimestampOffset firstEntry = idx.entry(0);
+        idx.sanityCheck();
+        assertDoesNotThrow(() -> idx.close());
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TimeIndexTest.java]: please remove `assertDoesNotThrow`

File: storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java [hunk 14]
```diff
@@ -0,0 +1,207 @@
...
+
+        assertDoesNotThrow(testIndex::flush);
```
Comment (chia7712) [inline storage/src/test/java/org/apache/kafka/storage/internals/log/TransactionIndexTest.java]: ditto, please remove those `assertDoesNotThrow`

**Comment by FrankYang0529** [comment]
Hi @chia7712, I addressed all comments. May you take a look? Thank you.

**Pull request merged**